marine.c

```diff
@@ -579,34 +579,37 @@ WS_DLL_PUBLIC marine_result *marine_dissect_packet(int filter_id, unsigned char
     return result;
 }
 
-
-WS_DLL_PUBLIC int compile_bpf(char *bpf, struct bpf_program *fcode, char *err_msg) {
+int compile_bpf(char *bpf, struct bpf_program *fcode, char *err_msg) {
     pcap_t *pc;
 
     pc = pcap_open_dead(DLT_EN10MB, MIN_PACKET_SIZE);
-    if (pc != NULL) {
-        struct bpf_program temp_fcode;
-        struct bpf_program *used_fcode;
-        int ret_value;
-
-        used_fcode = (fcode != NULL) ? fcode : &temp_fcode;
-        ret_value = pcap_compile(pc, used_fcode, bpf, 0, 0);
-
-        if (ret_value == -1) {
-             strcpy(err_msg, "Failed compiling the BPF");
-        }
-
-        if (fcode == NULL && ret_value == 0) {
-            pcap_freecode(used_fcode);
+    if (pc == NULL) {
+        return -1;
+    }
+    
+    if (pcap_compile(pc, fcode, bpf, 0, 0) == -1) {
+        if (err_msg != NULL) {
+            strcpy(err_msg, "Failed compiling the BPF");
         }
-
+        
         pcap_close(pc);
-        return ret_value;
+        return -1;
     }
 
-    return -1;
+    pcap_close(pc);
+    return 0;
 }
 
+WS_DLL_PUBLIC int validate_bpf(char *bpf) {
```

Comment:
This should return a boolean (i.e 0 and 1) instead of error codes, since there are no actual cases for different error codes :)